src/prediction/runner.rs

```diff
@@ -86,36 +99,104 @@ impl<'a, E: AiEnv> PromptRunner<'a, E> {
+                };
+                debug!("Setting flagstate to: {:?}", flagstate);
+                newflag.state = flagstate;
+                newflag.update(pool).await?;
+            }
+            None => {
+                let flagstate = match num_stmts {
+                    1 => StatementFlagState::Flagged,
+                    _ => StatementFlagState::MaybeFlagged,
+                };
+                debug!("Setting flagstate to: {:?}", flagstate);
+                StatementFlag::create(pool, stmt.id, flagstate, FlagCategoryContainer::Empty)
+                    .await?;
+            }
+        }
+    }
+    Ok(())
+}
 
-    use propolis_datas::{
-        apikey::{ApiKey, TransientApiKey},
-        statement::{FlagCategoryContainer, StatementFlagState},
-    };
+/// Used to select next key to use for requests
+pub struct ApiKeySelector {
+    /// Mapping of raw key to ApiKey instance
+    pub keys: HashMap<String, ApiKey>,
+}
 
-    let mut keys: HashMap<String, ApiKey> = HashMap::new();
-    let mut raw_keys: Vec<String> = opts.openai_api_keys;
-    if let Some(rk) = opts.openai_api_key {
-        raw_keys.push(rk);
+impl ApiKeySelector {
+    /// Collect additional environment keys of form OPENAI_API_KEY_n
+    /// Tries n ∈ {0, ..} until the first env var that can't be found
+    fn collect_numbered_openai_env_keys() -> Vec<String> {
+        let mut result: Vec<String> = vec![];
+        let mut n = 0;
+        while let Ok(raw_key) = std::env::var(format!("OPENAI_API_KEY_{n}")) {
+            result.push(raw_key);
+            n += 1;
+        }
+        result
     }
-    for raw_key in raw_keys {
-        let rkey = TransientApiKey::Raw(raw_key.to_owned());
-        let key = ApiKey::get_or_create(pool, &rkey, None::<String>)
-            .await
-            .expect("Unable to get_or_create key.");
-        keys.insert(raw_key, key);
+
+    /// Create an instance, creating keys in the DB (for stats only) if necessary
+    pub async fn create(opts: &PredictionOpts, pool: &mut SqlitePool) -> anyhow::Result<Self> {
+        let mut keys: HashMap<String, ApiKey> = HashMap::new();
+        let mut raw_keys: Vec<String> = opts.openai_api_keys.to_owned();
+        raw_keys.append(Self::collect_numbered_openai_env_keys().as_mut());
+        if let Some(rk) = &opts.openai_api_key {
+            raw_keys.push(rk.into());
+        }
+        for raw_key in raw_keys {
+            let rkey = TransientApiKey::Raw(raw_key.to_owned());
+            let key = ApiKey::get_or_create(pool, &rkey, None::<String>).await?;
+            keys.insert(raw_key, key);
+        }
+        Ok(Self { keys })
+    }
+
+    /// Yield next key to use
+    pub fn next(&self) -> anyhow::Result<(String, ApiKey)> {
+        let raw_key = self
+            .keys
+            .keys()
+            .collect::<Vec<&String>>()
+            .choose(&mut rand::thread_rng())
+            .ok_or(anyhow!("Unable to select any API key."))?
```

Comment:
Would be good to fail on program start when no API key is available, instead of later on usage.